src/Workspaces/Core/Portable/Utilities/ComparerWithState.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Immutable;
+
+namespace Roslyn.Utilities
+{
+    internal delegate int ComparisonWithState<in T, in S>(T x, T y, S state);
+
+    /// <summary>
+    /// Extension of <see cref="Comparer"/> to take into account dynamic <typeparamref name="S"/>
+    /// </summary>
+    /// <typeparam name="T">Comparing type</typeparam>
+    /// <typeparam name="S">State</typeparam>
+    internal class ComparerWithState<T, S>
```

Comment:
not certain if these classes are needed?  can't you make one bit ComparisonWithState out of N individual ones?